### DISCUSSION TITLE
Loading data on the frontend

### DISCUSSION THREAD

**Discussion by pavish** [mathesar-foundation/mathesar#514] [Maintainer discussions]
We need to decide upon the spreadsheet loading functionality, for the stability of CRUD operations on the frontend.

## Some context:
Our current spreadsheet loading works the following way:
* Data is fetched initially for the set of visible rows, along with a few additional buffer rows. Once fetched, that data is cached on the frontend.
* When user scrolls to a particular location, we identify the rows that are not loaded yet on the visible portion and fetch data for those rows.
* When user scrolls back to a location where data was already fetched, we do not fetch again.

When sort/group/filter is applied/cleared:
* We clear all existing rows, and refetch again for visible portion.
* When user scrolls, the above mentioned steps happen as usual.

Grouping
* When a group is applied, we make the grouped column as the main sort column.
* When iterating rows, we identify if the values of previous and current rows are different for grouped columns, and based on that we add a group header.

The problems with the current approach:

* There is no guarantee that the rows would remain in the same order in the db.
* Rows may be duplicated on the frontend, if there was a change in the db which changes the row order.
  - This may return rows that were already visible, when we scroll to a different position.

The plan I had in mind to resolve this issue:
* Always sort by primary key. Even when grouping/sort is applied, sort additionally by primary key.
* Identify and merge rows based on primary key, if both of the duplicated rows are visible to user.
* If duplicate rows are not visible to user, keep the rows that are visible, and clear the cache of rows that are not visible.

Considering this in combination with other operations, it becomes more complex:

Adding rows:
* When we add rows, we do not know where the row would end up. The UX we've discussed would work for a paginated table (because we would have already fetched everything that is displayed), but it will not work for a spreadsheet view. We cannot keep the newly created row in it's position until an explicit refresh is done by user. When the user scrolls, the row will change position/ be duplicated.
* We could mention a message stating that the row's position will be updated on scroll/on focusing out of the row, but where the row is placed cannot be indicated to the user, because the client may not have the data.

Editing rows:
* When a cell is edited, the row's position may change. Even with all the additional steps taken, a pk column could still be edited leading to a change in position on the db.
* This has the same drawback of row being placed on a different position upon scroll. When a filter/sort is applied, it would be clear to the user that the row's position may change, but when there is no such filter/sort, it will become hard for the user to edit a row and scroll down/up to check something else and come back to the row being edited.
* Keyboard based navigation would become unreliable, since the row's position is not absolute.

Deleting rows:
* Currently when we delete a row, we reload all visible rows. This removes the deleted row from the view.
* This works well if the order remains the same (which it usually would). However a change in order would change all the visible rows, and confuse the user. 

All the above operations become more complex when done in combination with grouping. This affects user experience and also implementation, as it is more error prone. Also, if there are changes to the data from a different client, row positions will keep changing, making it harder to the user to work with the spreadsheet.

The major reason for all these issues is that we cannot block user scroll action, when there is an update happening, and our record fetching depends entirely on the scroll action.

## A different approach

A more direct approach (used by airtable, google sheets)
* Fetch all the records on initial load.
* Sync all changes between db and the ui.

Here, a change in order would not be an issue, since all the records are fetched at once, so we never have to refetch data. Any change in db would be notified to the ui, which will make the changes on the rows on their position on the frontend (which would remain the same for the entire duration of the user accessing the spreadsheet).

However upon filter, sort and grouping, we would have to fetch everything again or let the ui perform the actions. This is not a good idea, since we would want the db to have control over this.

We could overcome this by requesting the filtered/sorted/grouped rows and only fetching the pks for all those rows. Since we have the data on hand, we can use the pks from the result to filter/order the rows accordingly. We send only the pks to keep the response smaller.

Cons:
* It would affect loading records for larger tables. We would have to limit the number of visible records to a considerable number, and combine it with pagination.

I was not in favour of this approach initially due to having to load all the records on the frontend, but for our current requirements, this seems to be the only viable option. I'm open to more suggestions.

---
All the above are not applicable for tables without pk, since they would only be displayed in a paginated table, and update and delete operations would not be possible over them, in a spreadsheet style.

**Comment by pavish** [comment]
There are also a set of product questions I have, related to the user experience and implementation.

While we want to make databases accessible, we are actually trying to provide a spreadsheet solution. In what levels/priorities do we mix these two together?

### Requirements for the two:

For a spreadsheet: (Airtable, google sheets, baserow)
  - We will always need a single unique key column, which is non-editable and is not displayed to the user.
  - The order needs to stay absolute. It should not change unless a sort option is specified. i.e order should be reliable.

To support a database: (NocoDB)
  - The ui has to be always paginated, true to native db functionality.

---
Apart from this, here are some initial level thoughts for the features on MVP. We need to plan the approaches of this out in advance, since they also affect how we show the data to the user:

### Requirements for collaboration:
  - Primary/Unique key that is used to identify the row has to be non-editable.
  - If we allow editing it, we would need to keep track of the original pk, which poses more questions, such as for how long/on what conditions.

### Requirements for undo/redo:
  - Detailed steps to identify a row, for which a pk is changed, multiple times by multiple clients.
    - Since we audit it, we should be able to identify based on audit information.
    - But if there are two Mathesar clients in parallel, undo on a client should only undo the changes made by that specific client.
    - This needs more thought, when both clients are editing the same row.

**Comment by kgodey** [comment]
There are a bunch of issues here, Since the posts above contain a lot of detail, I'll try to summarize them before responding to them. Please let me know if I've missed something.

## Summary

### Technical issues with infinite scroll

1. If we allow the user to edit primary keys, then the position of the row will need to move.
1. If user scrolling is used as a trigger to fetch data from the backend, then new rows being added or edited by other database clients (including other instances of Mathesar) will cause row position to change when the user scrolls back and forth. This is especially a problem if the user was in the middle of editing a row.
1. With infinite scroll, we can't keep added rows visible until the user navigates to a different page because there is no different page in the same table.
1. Generally, we don't have a way of distinguishing newly added rows to the user if they weren't there the last time they scrolled.
1. We need to ensure that whatever solution we use will work with collaboration and undo/redo

### Suggested solutions
- Don't show primary keys to the user
- Make primary keys non-editable from the UI 
- Either:
    - Load all data in the frontend
    - Load no data in the frontend and always sync with the backend when the user scrolls
- Implement WebSockets right away and switch the frontend to using them instead of the REST API

### Product and planning questions 
- What features of a spreadsheet do we actually want to provide to users?\

## Response
Given the issues I've listed above, here's what I think:

I don't think of Mathesar as a spreadsheet. Our main objective is to reduce the effort (friction) that a user needs to perform common tasks, and I think we should take some inspiration from spreadsheets; especially with editing in place (instead of requiring a form to edit a record) and easy bulk-editing operations. 

Our UI does have similarities to spreadsheet, but I'm concerned that thinking of Mathesar AS a spreadsheet might cause us to lose focus on what matters to users. That is, I'm not sure that infinite scroll is something that matters to Mathesar users, even though it's a standard feature of spreadsheets to have infinite scroll.

It seems like removing infinite scroll and using pagination for now will solve most of our problems:
- Our previously designs for editing and adding records will work.
- We won't be fetching data on scroll, we'll only be fetching data when the user explicitly wants to see a different set of data.
- We won't need to load a ton of data on the frontend (as we would if we loaded all data on the frontend like Airtable or Google Sheets)
- We won't need to divert backend time to implementing WebSocket-based APIs.
- The user won't expect their data to update in real-time.
- We can even support editing primary keys with no ill-effects.

I don't see any significant cons to losing infinite scroll. When people are working with data, they're not working with thousands of rows of data at once in their active memory, they're going to be filtering, sorting, grouping, etc. so they see the data they need.

### Recommendation
Given that our goal is to ship a usable MVP quickly, my recommendation is that we go back to using pagination for all tables. I don't think infinite scroll is a feature that will distinguish us from a usability perspective and it comes with technical complications that I don't think would be a good use of time for us to focus on solving.

- After we ship our MVP, if we find that enough users want infinite scroll, we can figure out how to implement it.
- We can spec out collaboration and undo/redo when we get to them. I'd like to take the approach of completing one feature at a time and we'd have to pause design and development work on other features to think about and spec these out.

### Questions
- What should our default page size be? 
  - I was thinking 500 records – it should keep the data loaded on the page small.
  - We should, of course, allow people to change the page size if desired.

**Reply by pavish** [reply]
Here are some of my thoughts:

> We won't need to divert backend time to implementing WebSocket-based APIs

We will need websockets for collaboration, and we will have to spend time to build it eventually before the MVP.

> We can even support editing primary keys with no ill-effects.

I'm not sure how well we can support this, when collaboration and undo/redo are involved.

> I don't see any significant cons to losing infinite scroll.
* Grouping would not be an effective feature.
* Adding new rows when pagination is involved, is not intuitive and hard to understand.
* We lose good user experience, especially for non-technical users.
* We lose what could be a good selling point for Mathesar.

### My recommendation:
* Let's set the default page size at a higher limit, around 5000.
  * Above 5000 rows, it will be paginated. It will work with pagination + virtual scroll (so data doesn't change but rendered in chunks, so UI would be fast).
  * This should give a good spreadsheet experience for a higher number of users, compared to 500.
  * We will need websockets, for collaboration. So, I think it might be worth it to implement it early.
  * Suggestion: We can read records as a stream from postgres, and send it as small batches to the open websocket line, as we read it from db.

**Reply by kgodey** [reply]
@pavish I broke up my response into sections for readability.

### WebSockets & Primary Keys

> We will need websockets for collaboration, and we will have to spend time to build it eventually before the MVP.

That's our plan, yes, but I think it's best to build things out in an incremental way so that Mathesar becomes progressively more usable over time. In the worst case scenario, if a bunch of things go over schedule, we can ship Mathesar's MVP without needing to build real-time collaboration (or WebSocket APIs).

However, Mathesar becomes infinitely less useful without data types or views or data modeling, which is what the backend is scheduled to spend time on in the next few weeks. Mathesar is also not useful if we cannot edit or add records. I'd rather get these things completed first rather than switching to spend a bunch of time on WebSockets (which we can do without if needed) and delay everything else.

> I'm not sure how well we can support this, when collaboration and undo/redo are involved.

Good point. It's probably safest to make primary keys non-editable now and figure out how to make them editable later.

### Cons of Infinite Scroll

> Grouping would not be an effective feature.

I'm not sure what you mean by this. How is grouping not an effective feature and what's the relationship to infinite scroll?

> Adding new rows when pagination is involved, is not intuitive and hard to understand.

I'm also sure what you mean by this. How is adding new rows hard to understand and how would that be better than infinite scroll?

> We lose good user experience, especially for non-technical users.

Again, I'm not sure what "good user experience" is or the relationship is to infinite scroll.

> We lose what could be a good selling point for Mathesar.

I assume this is based on what you were saying in our call about how there aren't database GUI products with infinite scroll available. It seems like a major assumption that infinite scroll is a selling point.

In general, all of the cons you listed are very broad assertions that make big assumptions about what users would want. In addition, even if there are specific user experience concerns, it's another big assumption that infinite scroll is the only solution to that design problem.

### Recommendation

I've addressed the WebSockets portion already.

My concern with a default page size of 5000 is that it stores a lot of data on the client and will result in degraded performance for users with slow internet connections or slow computers, for what seems like no good reason. Relatedly, this recommendation does not address _why_ a user would be better served with seeing so many results in a page.

**Reply by pavish** [reply]
https://github.com/centerofci/mathesar/discussions/514#discussioncomment-1136630

**Comment by pavish** [comment]
@kgodey The cons I've mentioned above compare infinite scroll to our design specs. I am not completely on the side of infinite scroll, my point is that it offers a better usability to me personally, compared to our current pagination based design spec.

---

> How is grouping not an effective feature and what's the relationship to infinite scroll?

With infinite scroll:
* The user has the potential option to minimize groups.
* The user could view all group information with counts, in one place (when minimized), and have a complete overview.

These are not possible with our current pagination based UX.

Suggestion:

We can probably think of a different UX for grouping, which provides these options for a paginated table:
* Instead of showing a group header among the rows, what if we have a horizontal section with the group information, which acts like a filter?
  * Like the outline view of a PPT.
  * Here's a mockup.
  * ![Group 32](https://user-images.githubusercontent.com/11850603/128423781-e9b66319-374a-44c6-affd-bcfa9944cc71.png)
  * Ofcourse, this still has it's drawbacks, where it will have to load a lot of data, if the user groups by a unique column (Not sure why anyone would do that, but it's a scenario).

* We can experiment and come up with better design options.

---

> How is adding new rows hard to understand and how would that be better than infinite scroll?

With infinite scroll:
* Rows are always added at the end. This is a natural flow of an user.

With paginated table:
* Rows addition can be done at any page, and it will change position on refresh or page change. While this is indicated with a message, the user cannot add/edit a row and move to a different page and come back to find the same row.
* This is not a major issue, but an annoyance, especially if the user wants to compare rows while editing.

Suggestion:
* We can think of a UX, where a row that the user is currently adding/editing, is always visible even when the user is changing pages. I'm not able to think of an UX right out of my mind, but the concern is that it might be better if we allow users to change pages, while attempting to edit a particular row, and keeping that row in view.
* This is, as I mentioned, an annoyance more than a requirement, but definitely something we should keep in mind.

---

> Again, I'm not sure what "good user experience" is or the relationship is to infinite scroll.

* Non-technical users are more used to spreadsheet views with infinite scroll. This is mainly due to the wide range of spreadsheet softwares available.
* By good user experience, I meant providing users the interface that they are well adjusted to.
* But the major concern I had was to provide a more intiutive UX, like with grouping.

---

> It seems like a major assumption that infinite scroll is a selling point.

* I agree that this is a major assumption. This stemmed from my thoughts that tie to the above points on good user experience.

---

> This recommendation does not address why a user would be better served with seeing so many results in a page

To make it simpler for new users, it is best to provide an interface they are used to.
* We could come up with more creative solutions to work with pagination and satisfy needs such as viewing all groups (which I do assume would be the major reason for anyone wanting to group by columns).
* But these solutions would be new and not trial tested, with respect to usability. A spreadsheet-like interface, with a larger page size, would retain the spreadsheet experience, and users wouldn't have to take time figuring out stuff on the UI.

While this is mostly a non-issue, this is a valid user experience concern.

As mentioned above, I'm not fixated to go with infinite scroll, but rather not completely satisfied with our current pagination UX.

--- 

I agree to all the points on not prioritizing websockets right away.

**Reply by ghislaineguerin** [reply]
@pavish I think your grouping idea makes sense and we should try creating a prototype for it. I also agree that the current group UI would not work with pagination.

**Reply by kgodey** [reply]
@pavish Thanks for the explanations.

Generally, it sounds like you have a personal preference for the user experience of infinite scroll and it seems like many of the points you made are based on that preference. I don't think we should assume that that preference will be true for all/most users (it's not true for me, for example). 

Since we don't know which experience will be best for users or if there's a notable benefit to infinite scroll, let's go with the simpler solution for now, which is pagination for all tables for the MVP (with 500 records). Based on user input, we can decide on whether to prioritize infinite scroll or larger page sizes. This decision also follows our [product process principles](https://wiki.mathesar.org/product/principles) of keeping implementation simple and iterating.

Regarding grouping and adding rows, your concerns and ideas make sense to me. For now, I think we should implement the current UX from the design specs as-is and revisit it once we have basic functionality deployed to our staging server and we can play around with real data. I think that'll give us better ideas for how to make the experience better.

We can revisit grouping UX as part of [16. Summarized Data](https://github.com/centerofci/mathesar/milestone/16) (which revolves entirely around groups), and improving adding/editing as part of [17. Bulk Operations on Records](https://github.com/centerofci/mathesar/milestone/9) and [18. Spreadsheet-like Editing](https://github.com/centerofci/mathesar/milestone/10) (both of which focus on the CRUD experience for records). @ghislaineguerin, does this make sense to you?

---

I assume this unblocks you for:
- #195 
- #197

Let me know if you have any other blockers to implementation.

**Reply by pavish** [reply]
@kgodey Sounds good to me!

**Reply by ghislaineguerin** [reply]
@kgodey sounds good to me
